Platform/OpenNtfsDxe/Data.c

```diff
@@ -0,0 +1,566 @@
+      return EFI_DEVICE_ERROR;
+    }
+
+    Cluster = Cluster * mClusterSize;
+
+    if (Index == (ClustersTotal - 1)) {
+      Size = (UINTN) ((Length + Offset) & (mClusterSize - 1));
+
+      if (Size == 0) {
+        Size = mClusterSize;
+      }
+
+      if (Index != Runlist->TargetVcn) {
+        OffsetInsideCluster = 0;
+      }
+    }
+
+    if (Index == Runlist->TargetVcn) {
+      Size -= (UINTN) OffsetInsideCluster;
+    }
+
+    if ((Index != Runlist->TargetVcn) && (Index != (ClustersTotal - 1))) {
+      Size = mClusterSize;
+      OffsetInsideCluster = 0;
+    }
+
+    if (Cluster != 0) {
+      Status = DiskRead (
+        Runlist->Unit.FileSystem,
+        Cluster + OffsetInsideCluster,
+        Size,
+        Dest
+        );
+      if (EFI_ERROR(Status)) {
+        return Status;
+      }
+    } else {
+      SetMem (Dest, Size, 0);
+    }
+
+    Dest += Size;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+DiskRead (
+  IN EFI_FS   *FileSystem,
+  IN UINT64   Offset,
+  IN UINTN    Size,
+  IN OUT VOID *Buffer
+  )
+{
+  EFI_STATUS         Status;
+  EFI_BLOCK_IO_MEDIA *Media;
+
+  ASSERT(FileSystem != NULL);
+  ASSERT(FileSystem->DiskIo != NULL);
+  ASSERT(FileSystem->BlockIo != NULL);
+
+  Media = FileSystem->BlockIo->Media;
+
+  Status = FileSystem->DiskIo->ReadDisk (
+    FileSystem->DiskIo,
+    Media->MediaId,
+    Offset,
+    Size,
+    Buffer
+    );
+  if (EFI_ERROR(Status)) {
+    DEBUG((DEBUG_INFO, "NTFS: Could not read disk at address %08x\n", Offset));
+    return Status;
+  }
+
+  return EFI_SUCCESS;
+}
+
+EFI_STATUS
+EFIAPI
+ReadMftRecord (
+  IN  EFI_NTFS_FILE *File,
+  OUT UINT8         *Buffer,
+  IN  UINT64        RecordNumber
+  )
+{
+  EFI_STATUS     Status;
+
+  Status = ReadAttr (
+    &File->MftFile.Attr,
+    Buffer,
+    RecordNumber * mFileRecordSize,
+    mFileRecordSize
+    );
+  if (EFI_ERROR(Status)) {
+    DEBUG((DEBUG_INFO, "NTFS: Could not read MFT Record 0x%Lx\n", RecordNumber));
+    return Status;
+  }
+
+  return Fixup (Buffer, mFileRecordSize, "FILE");
+}
+
+EFI_STATUS
+EFIAPI
+ReadAttr (
+  IN  NTFS_ATTR *Attr,
+  OUT UINT8     *Dest,
+  IN  UINT64    Offset,
+  IN  UINTN     Length
+  )
+{
+  EFI_STATUS       Status;
+  UINT8            *Current;
+  UINT32           Type;
+  UINT8            *AttrStart;
+  UINT64           Vcn;
+  ATTR_LIST_RECORD *Record;
+  ATTR_HEADER_RES  *Res;
+
+  Current = Attr->Current;
+  Attr->Next = Attr->Current;
+
+  if ((Attr->Flags & NTFS_AF_ALST) != 0) {
+    if (Attr->Last < (Attr->Next + sizeof (ATTR_LIST_RECORD))) {
+      DEBUG((DEBUG_INFO, "NTFS: $ATTRIBUTE_LIST does not contain even a single record.\n"));
+      return EFI_VOLUME_CORRUPTED;
+    }
+
+    Record = (ATTR_LIST_RECORD *) Attr->Next;
+    Type = Record->Type;
+
+    Vcn = DivU64x64Remainder (Offset, mClusterSize, NULL);
+    if (COMPRESSION_BLOCK >= mClusterSize) {
+      Vcn &= ~0xFULL;
+    }
+
+    Record = (ATTR_LIST_RECORD *) ((UINT8 *)Record + Record->RecordLength);
```

Comment:
No need to cast to `(ATTR_LIST_RECORD *)`

Reply:
Cast is mandatory, as sizeof (UINT8) < sizeof (ATTR_LIST_RECORD).